File: Introibo/Data/Saint.swift
import Foundation

// Matches Introibo/Resources/saints.json.
// Ported from prototype/saints.html's SAINTS object.

struct Saint: Identifiable, Decodable, Hashable {
    let slug: String
    var name: String
    var title: String
    var quote: String
    var penance: String?
    let penanceLatin: String?
    var sections: [Section]
    var prayers: [SaintPrayer]?
    var related: [RelatedLink]? = nil

    var id: String { slug }

    struct Section: Decodable, Hashable {
        let lat: String
        var eng: String
        var practices: [Practice]
    }

    struct Practice: Decodable, Hashable {
        var t: String
        var d: String
    }

    struct SaintPrayer: Decodable, Hashable, Identifiable {
        var title: String
        let latin: String?
        var eng: String
        var note: String?

        var id: String { title }
    }
}

File: android/app/src/main/java/com/lampstandhq/introibo/data/model/Saint.kt
package com.lampstandhq.introibo.data.model

import com.lampstandhq.introibo.data.links.RelatedLink
import kotlinx.serialization.Serializable

/**
 * Matches saints.json.
 */
@Serializable
data class Saint(
    val slug: String,
    val name: String,
    val title: String,
    val quote: String,
    val penance: String? = null,
    val penanceLatin: String? = null,
    val sections: List<Section>,
    val prayers: List<SaintPrayer>? = null,
    val related: List<RelatedLink>? = null,
) {
    @Serializable
    data class Section(
        val lat: String,
        val eng: String,
        val practices: List<Practice>,
    )

    @Serializable
    data class Practice(
        val t: String,
        val d: String,
    )

    @Serializable
    data class SaintPrayer(
        val title: String,
        val latin: String? = null,
        val eng: String,
        val note: String? = null,
    )
}
